### PR TITLE
tests: Improve k8s negative tests

### PR DESCRIPTION
**Pull request by BbolroC**
This PR improves the k8s negative tests by:

- Making the `kubectl` commands in `print_node_journal()` synchronous, similar to the fix for `exec_host()`, to prevent missing journal logs due to premature termination of the node debugger pod.
- Completing the tests as soon as a specified log message is captured, rather than waiting for the default 2 minutes to confirm that the test pod is not running.

For a detailed explanation, please refer to the commit messages.

Signed-off-by: Hyounggyu Choi <Hyounggyu.Choi@ibm.com>

### CODE REVIEW

**Review by amshinde — APPROVED** [review]

**Comment by BbolroC** [comment]
This PR will be improved when #10232 is merged. Thanks!

**Review by GabyCT — APPROVED** [review]

**Comment by BbolroC** [comment]
# Please do not merge this until I got final confirmation from Wainer. Thanks!

**Review by ChengyuZhu6 — APPROVED** [review]
LGTM thanks @BbolroC

File: tests/integration/kubernetes/lib.sh
Comment (wainersm) [inline tests/integration/kubernetes/lib.sh:254]: Hi @BbolroC !

Never occurred to me that the debugger pod could end prematurely. Good catch!

File: tests/integration/kubernetes/lib.sh
Comment (wainersm) [inline tests/integration/kubernetes/lib.sh:308]: I'm wondering why we never used the `exec_host()` here.... IIRC `print_node_journal()` was copied from CCv0 branch and `exec_host()` already existed on main (but not CCv0). Maybe we overlooked `exec_host()`?

I don't see any technical impeachment for using exec_host, expect that maybe we would need to move exec_host from tests_common.sh to lib.sh to avoid any cyclical dependency between the two libs. @BbolroC does it sound as a good idea to you?  --- Can be done in a separate PR.

File: tests/integration/kubernetes/lib.sh
Comment (wainersm) [inline tests/integration/kubernetes/lib.sh:136]: Hi @BbolroC ,

Suppose that check_logs() returned true after 30min then it will hit this line and wait  `assert_pod_fail`. The time for this test to finish is still 2min. am I missing something?

If I'm right, I was thinking in implement something like:
* do not run assert_pod_fail (not even on background)
* run check_logs
  * on check_logs true, check one-time pod is not Ready
  * on check_logs false, the test failed and it will return immediately

One warn for my idea: it's too fragile. Any bug on `check_logs` will likely to make the test always to pass (check_logs returns and the one-time check passes too because the pod didn't have time to be Ready, but it might get ready later).

File: tests/integration/kubernetes/lib.sh
Comment (BbolroC) [reply tests/integration/kubernetes/lib.sh:254]: Yeah, this happens frequently on s390x (even yesterday. see https://github.com/kata-containers/kata-containers/actions/runs/10981530559/job/30489117743, with two false negatives). 

I initially thought that it was an environmental issue on the platform. But after looking at #10081, I realised that it could be related to how we are handling `kubectl debug` in our tests. 

My goal is to run a dedicated `custom-node-debugger` pod first, which was set up in #10232, and use `kubectl exec` on the pod until the test is complete. 

I'm hoping this will resolve the issues around `kubectl debug`. 😉

File: tests/integration/kubernetes/lib.sh
Comment (BbolroC) [reply tests/integration/kubernetes/lib.sh:308]: > This PR will be improved when https://github.com/kata-containers/kata-containers/pull/10232 is merged. Thanks!

Yes, that is why I mentioned it earlier. Once this PR is merged, all `kubectl debug` will be replaced with `exec_host()`, giving us a single point of control for that functionality. Thanks.

File: tests/integration/kubernetes/lib.sh
Comment (BbolroC) [reply tests/integration/kubernetes/lib.sh:136]: Yes and No:

Yes: the background job logic does follow your suggestion, `assert_pod_fail()` reaches the `wait` immediately after https://github.com/kata-containers/kata-containers/blob/8a893cd4ee274d3f457fc13c2aaabb6532211f01/tests/integration/kubernetes/lib.sh#L52 finishes. At that point, the test depends solely on `check_logs()`. If the message is captured and the pod isn’t in a running state, the test passes. Otherwise, it fails if the message does not appear within 120s (the default duration for `check_logs()`). But, of course, we can make `assert_logs_pod_fail()` synchronous as suggested unless `assert_pod_fail()` is used - which incurs an unnecessary 2-min wait due to `k8s_create_pod()`.

No: I don't think there is any chance a pod that produces an error message would end up in a ready state (even the negative test is not designed to have a pod ready) But you are saying the test should check that an error message is captured and ensure a pod doesn't get ready at least within 2 mins. Then yes, keeping the current logic would be better. 😉 

If you are still concerned about a background job, I can change the logic as you suggested. Just say the word, my lord. Thanks.

File: tests/integration/kubernetes/lib.sh
Comment (stevenhorsman) [inline tests/integration/kubernetes/lib.sh:1]: Do you think we improve the code logic of `assert_pod_fail` to check for an error state with polling rather than the 2min wait inherited from `k8s_create_pod`, which as you point out is a waste of time often, or do you think that approach would be too fragile? As it stands I'm not sure that `assert_pod_fail` is necessarily fit for purpose as it seems to rely on pid checking in your new code, which is non-trivial for other users of it?

I guess it might have similar co

**Comment by BbolroC** [comment]
Thanks to @stevenhorsman feedback, I will update the code to follow:

- Leveraging the `custom-node-debugger` to replacing all `kubectl debug`
- For negative tests,
    - First, asserting a pod to fall into a failed state with `kubectl wait --for=condition=failed`
    - Finally, checking the host journal if a message of interest appears

**Comment by stevenhorsman** [comment]
> Thanks to @stevenhorsman's feedback, I will update the code to follow:
> 
> * Leveraging `exec_host()` to replacing all `kubectl debug`
> * For negative tests,
>   
>   * First, asserting a pod to fall into a failed state with `kubectl wait --for=condition=failed`
>   * Finally, checking the host journal if a message of interest appears

That sounds great. I think if we have leave the existing calls to `assert_pod_fail` and `assert_logs_contain` as is in the tests, but use your great ideas to make the functions much smarter in `lib.sh` then hopefully it is still easy for all callers to understand, but will work quicker and with less errors. Thanks @BbolroC!

**Comment by BbolroC** [comment]
Why polling with `$(kubectl get pod "${pod_name}" -o jsonpath='{.status.containerStatuses[0].state.waiting.reason}') = *BackOff*` is used instead of `kubectl wait --for=condition=Failed pod/${pod_name} --timeout=${timeout}`:

- I prefer not to use `${timeout}` because it forces a fixed wait time, which often has to be set quite generously.

**Comment by BbolroC** [comment]
Running time improvement (comparison with one sample):

| Name | Before | After |
|-------|--------|------|
| run-k8s-tests-on-zvsi(nydus, qemu-coco-dev, k3s) | 39m | 25m |
| run-k8s-tests-coco-nontee(qemu-coco-dev, nydus, guest-pull) | 53m | 42m |
| run-k8s-tests-on-tdx(qemu-tdx, nydus, guest-pull) | 59m | 29m |
| run-k8s-tests-on-sev(qemu-sev, nydus, guest-pull) | 36m | 31m || 

`run-k8s-tests-sev-snp(qemu-snp, nydus, guest-pull)` failed due to timeout, which does not look relevant to this change.

**Review by stevenhorsman — APPROVED** [review]
This looks like a great improvement and hopefully should resolve some of the reliability of the `assert_logs_contains` usages as well as improve the test run time. Thanks very much @BbolroC!

File: tests/integration/kubernetes/k8s-confidential-attestation.bats
Comment (BbolroC) [reply tests/integration/kubernetes/k8s-confidential-attestation.bats:99]: Yeah, I am preparing another improvement PR after this one (assuming it gets merged). I will include this deduplication in that. Thanks. 😉

**Comment by stevenhorsman** [comment]
> # Please do not merge this until I got final confirmation from Wainer. Thanks!

@fidencio 👀 😆

**Pull request merged**
